### PR TITLE
Migrate win_iis_webapppool module to new microsoft.iis repository

### PR DESCRIPTION
**Pull request by ronger4**
##### SUMMARY

1) Migrate win_iis_webapppool module to new microsoft.iis repository
2) Upgrade the module to the Ansible.Basic style module.
3) Name of the module changed to web_app_pool
4) Documentation updated
5) info module was added and called web_app_pool_info

##### ISSUE TYPE

Feature Pull Request

##### COMPONENT NAME

win_iis_webapppool
##### ADDITIONAL INFORMATION

```paste below
Things to give extra look in the review due to the acceptence criteria of the task:
- update the format. it currently has the old legacy module format which has poor logging and error handling
- support any diff output
- support for setting the advanced attributes of each object being managed
```

### CODE REVIEW

**Comment by jborean93** [comment]
> ansbile-lint : it supposed to ignore it no?

I didn't add the exclusion for the tests dir to like I did with https://github.com/ansible-collections/microsoft.iis/blob/main/.ansible-lint. It is up to you if you wish to do the same or want to lint these files. Personally I think we should exclude them and it isn't required to do so but if you are maintaining this going forward it'll be your choice.

> What is the best way to go forward here? migrate win_file_version to ansible.windows ?

To add more dependencies you will need to add the collection name to https://github.com/ansible-collections/microsoft.iis/blob/main/tests/requirements.yml. But for a simple task like this I probably would recommend you avoid depending on `community.windows` (the less deps the better). Looking at the test we can avoid the task altogether as IIS 8.0 is shipped with Server 2012 and we have a minimum of 2016 so the checks are no longer required. I'll add some comments to show what can be removed.

File: plugins/modules/web_app_pool.ps1 [hunk 1]
```diff
@@ -0,0 +1,347 @@
+#!powershell
+
+# Copyright: (c) 2015, Henrik Wallström <henrik@wallstroms.nu>
+# Copyright: (c) 2017, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#AnsibleRequires -CSharpUtil Ansible.Basic
+
+$ErrorActionPreference = 'Stop'
+
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:10]: Ansible always runs with this, we can remove this as it is redundant

```suggestion
```

File: plugins/modules/web_app_pool.ps1 [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        }
+    }
+    supports_check_mode = $true
+}
+$module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)
+
+$name = $module.Params.name
+$state = $module.Params.state
+$check_mode = $module.CheckMode
+$module.Result.info = @{
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:35]: I highly recommend we remove all these return values and don't return anything. This type of work should go into an `_info` module and not returning it here simplifies the code needed especially when it comes to check mode. The common practice for modules these days is to only return values that the user does not already know, for example a unique id of the resource they can use for references in subsequent tasks.

File: plugins/modules/web_app_pool.ps1 [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+    processModel = @{}
+    recycling = @{
+        periodicRestart = @{}
+    }
+}
+$module.Result.changed = $false
+
+# Stores the free form attributes for the module
+$attributes = $module.Params.attributes
+$module.Result.attributes = $attributes
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:50]: Ditto here, we should avoid returning what we already know, in this case the `attributes` are just the input values and they are automatically added by `Ansible.Basic` when running with a verbosity of `-vvv` or higher which is common Ansible practice.

File: plugins/modules/web_app_pool.ps1 [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+$pool = Get-Item -LiteralPath IIS:\AppPools\$name -ErrorAction SilentlyContinue
+if ($state -eq "absent") {
+    # Remove pool if present
+    if ($pool) {
+        try {
+            Remove-WebAppPool -Name $name -WhatIf:$check_mode
+        }
+        catch {
+            $module.FailJson("Failed to remove Web App pool $($name): $($_.Exception.Message)")
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:199]: A nice feature of `FailJson` is it accepts a second argument which we can use for providing better error messages. We just need to add `$_` as an argument and `FailJson` will do the rest for us.

```suggestion
            $module.FailJson("Failed to remove Web App pool $($name): $($_.Exception.Message)", $_)
```

An alternative is to just remove the `try/catch` altogether but sometimes it is nice to have a more descriptive error message.

File: plugins/modules/web_app_pool.ps1 [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+}
+else {
+    # Add pool if absent
+    if (-not $pool) {
+        if (-not $check_mode) {
+            try {
+                New-WebAppPool -Name $name > $null
+            }
+            catch {
+                $module.FailJson("Failed to create new Web App Pool $($name): $($_.Exception.Message)")
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:212]:
```suggestion
                $module.FailJson("Failed to create new Web App Pool $($name): $($_.Exception.Message)", $_)
```

File: plugins/modules/web_app_pool.ps1 [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+                if ($new_value -is [Array]) {
+                    try {
+                        Clear-ItemProperty -LiteralPath IIS:\AppPools\$name -Name $attribute_key -WhatIf:$check_mode
+                    }
+                    catch {
+                        $msg = -join @(
+                            "Failed to clear attribute to Web App Pool $name. Attribute: $attribute_key, "
+                            "Exception: $($_.Exception.Message)"
+                        )
+                        $module.FailJson($msg)
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:244]:
```suggestion
                        $module.FailJson($msg, $_)
```

File: plugins/modules/web_app_pool.ps1 [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+                    foreach ($value in $new_value) {
+                        try {
+                            New-ItemProperty -LiteralPath IIS:\AppPools\$name -Name $attribute_key -Value @{value = $value } -WhatIf:$check_mode > $null
+                        }
+                        catch {
+                            $msg = -join @(
+                                "Failed to add new attribute to Web App Pool $name. Attribute: $attribute_key, "
+                                "Value: $value, Exception: $($_.Exception.Message)"
+                            )
+                            $module.FailJson($msg)
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:255]:
```suggestion
                            $module.FailJson($msg, $_)
```

File: plugins/modules/web_app_pool.ps1 [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+                else {
+                    try {
+                        Set-ItemProperty -LiteralPath IIS:\AppPools\$name -Name $attribute_key -Value $new_value -WhatIf:$check_mode
+                    }
+                    catch {
+                        $msg = -join @(
+                            "Failed to set attribute to Web App Pool $name. Attribute: $attribute_key, "
+                            "Value: $new_value, Exception: $($_.Exception.Message)"
+                        )
+                        $module.FailJson($msg)
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:268]:
```suggestion
                        $module.FailJson($msg, $_)
```

File: plugins/modules/web_app_pool.ps1 [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+
+        # Set the state of the pool
+        if ($pool.State -eq "Stopped") {
+            if ($state -eq "started" -or $state -eq "restarted") {
+                if (-not $check_mode) {
+                    try {
+                        Start-WebAppPool -Name $name > $null
+                    }
+                    catch {
+                        $module.FailJson("Failed to start Web App Pool $($name): $($_.Exception.Message)")
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:283]:
```suggestion
                        $module.FailJson("Failed to start Web App Pool $($name): $($_.Exception.Message)", $_)
```

File: plugins/modules/web_app_pool.ps1 [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+            }
+        }
+        else {
+            if ($state -eq "stopped") {
+                if (-not $check_mode) {
+                    try {
+                        Stop-WebAppPool -Name $name > $null
+                    }
+                    catch {
+                        $module.FailJson("Failed to stop Web App Pool $($name): $($_.Exception.Message)")
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:296]:
```suggestion
                        $module.FailJson("Failed to stop Web App Pool $($name): $($_.Exception.Message)", $_)
```

File: plugins/modules/web_app_pool.ps1 [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+                }
+                $module.Result.changed = $true
+            }
+            elseif ($state -eq "restarted") {
+                if (-not $check_mode) {
+                    try {
+                        Restart-WebAppPool -Name $name > $null
+                    }
+                    catch {
+                        $module.FailJson("Failed to restart Web App Pool $($name): $($_.Exception.Message)")
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:307]:
```suggestion
                        $module.FailJson("Failed to restart Web App Pool $($name): $($_.Exception.Message)", $_)
```

File: plugins/modules/web_app_pool.ps1 [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+        }
+    }
+}
+
+# Manually get the periodicRestart attributes in recycling
+foreach ($attribute in $pool.recycling.periodicRestart.Attributes) {
+    $attribute_name = $attribute.Name
+    $attribute_value = $pool.recycling.periodicRestart.$attribute_name
+    $module.Result.info.recycling.periodicRestart.Add($attribute_name, $attribute_value)
+}
```
Comment (jborean93) [inline plugins/modules/web_app_pool.ps1:345]: If we decide to remove the return values in favour of a dedicated `_info` module then we can remove all this.

File: plugins/modules/web_app_pool.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2015, Henrik Wallström <henrik@wallstroms.nu>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+DOCUMENTATION = r'''
+---
+module: web_app_pool
+short_description: Configure IIS Web Application Pools
+version_added: 1.0.0
```
Comment (jborean93) [inline plugins/modules/web_app_pool.py:11]: Because this is the initial release we don't need to say when it was added.

```suggestion
```

File: tests/integration/targets/web_app_pool/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,43 @@
+---
+# Cannot use win_feature to install IIS on Server 2008.
+# Run a brief check and skip hosts that don't support
+# that operation
+- name: check if win_feature will work on test host
```
Comment (jborean93) [inline tests/integration/targets/web_app_pool/tasks/main.yml:5]: We can remove this check and the block as we now have a minimum versions of 2016 so `Get-WindowsFeature` is always present.

Basically just start with line 11 without the `when: module_available.rc == 0` check right at the end.

File: tests/integration/targets/web_app_pool/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+  - name: reboot after feature install
+    ansible.windows.win_reboot:
+    when: feature_install.reboot_required
+
+  - name: set version of IIS for tests
+    community.windows.win_file_version:
+      path: C:\Windows\System32\inetsrv\w3wp.exe
+    register: iis_version
+
```
Comment (jborean93) [inline tests/integration/targets/web_app_pool/tasks/main.yml:27]: As mentioned in https://github.com/ansible-collections/microsoft.iis/pull/6#issuecomment-2516144683 we should try and avoid adding extra dependencies to our tests, especially `community.windows`. Luckily in this case we don't need this anymore as it was used to see if IIS 8 was present and since our base Windows version is 2016 we can guarantee that.

```suggestion
```

File: tests/integration/targets/web_app_pool/tasks/tests.yml [hunk 16]
```diff
@@ -0,0 +1,423 @@
...
+      recycling.periodicRestart.schedule: [ "00:10:00", "10:10:00" ]
+  register: collection_change_again
+
+- name: assert results of set web pool attribute that is a collection (idempotent)
+  assert:
+    that:
+    - collection_change_again is not changed
+
+# The following tests are only for IIS versions 8.0 or newer
+- block:
```
Comment (jborean93) [inline tests/integration/targets/web_app_pool/tasks/tests.yml:357]: We can remove this block and the when on line 423 and always run these tests as IIS 8 will be present on Server 2016 and newer.

**Review by jborean93 — CHANGES_REQUESTED** [review]
Have added a few comments. Most of the stuff looks good, just a few minor suggestions really except for the return values.

I think we should take this opportunity of a clean break to avoid returning values we don't need to. That functionality should be reserved for a future `web_app_pool_info` module and keep this one dedicated to making the changes. This will result in easier to maintain code and making it align with common practice with Ansible modules.

File: tests/integration/targets/web_app_pool/tasks/main.yml (hunk 14 above)
Comment (ronger4) [reply tests/integration/targets/web_app_pool/tasks/main.yml:5]: There is an always statement to make sure the test pool is deleted so maybe worth keeping the block just for this case (without the when)

File: plugins/modules/web_app_pool.ps1 (hunk 1 above)
Comment (Yaish25491) [reply plugins/modules/web_app_pool.ps1:10]: @ronger4 Please split this module into a module and an _info

File: plugins/modules/web_app_pool.py (hunk 13 above)
Comment (Yaish25491) [reply plugins/modules/web_app_pool.py:11]: Please convert the documentation file to a ymal file

**Comment by ronger4** [comment]
I commit amended the fixes.
Few notes:
1) As info removed from result, many assertions were based on that in the integration tests.
2) I will add to this PR the *_info module as-well, after it will be ready, I could use the info module to fetch the app pool data from the system and use the data I get there to validate the main module in the integration test. (hope this is best practice)
3) I converted the .py file to be yml.
4) The result section in the doc file is empty now because I removed the info and attributes return as you suggested.

File: plugins/modules/web_app_pool.ps1 (hunk 1 above)
Comment (ronger4) [reply plugins/modules/web_app_pool.ps1]: Yes, will do
Work in progress

Comment (ronger4) [reply plugins/modules/web_app_pool.ps1]: > @ronger4 Please split this module into a module and an _info
Yes, will do
Work in progress

Comment (ronger4) [reply plugins/modules/web_app_pool.ps1]: > @ronger4 Please split this module into a module and an _info


Yes, will do
Work in progress

File: tests/integration/targets/web_app_pool/tasks/main.yml (hunk 14 above)
Comment (jborean93) [reply tests/integration/targets/web_app_pool/tasks/main.yml]: Yep sorry I got ahead of myself, I meant remove just the when.

**Comment by ronger4** [comment]
@jborean93 I add fixes according to your suggestions, added info module and utilized it in the tests.
Can you review please?
P.S: fails on lint but we can ignore it as agreed.

File: plugins/modules/web_app_pool_info.ps1 [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+    }
+    supports_check_mode = $true
+}
+
+$module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)
+
+$name = $module.Params.name
+
+$module.Result.exists = $false
+$module.Result.appPools = @()
```
Comment (jborean93) [inline plugins/modules/web_app_pool_info.ps1:20]: We try and keep the `snake_case` format for any input/output values in a module.

File: plugins/modules/web_app_pool_info.yml [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+
+EXAMPLES: |
+  - name: Return information about an existing application pool
+    microsoft.iis.web_app_pool_info:
+      name: DefaultAppPool
+    register: stored_info
+  
+  - name: Returns information about all application pools that exist on the system
+    microsoft.iis.web_app_pool_info:
+      name:
```
Comment (jborean93) [inline plugins/modules/web_app_pool_info.yml:34]: A better example might be to omit this option altogether rather than the `null` one you have here.

File: plugins/modules/web_app_pool_info.ps1 [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    $pool = Get-Item -LiteralPath IIS:\AppPools\$name
+    if ($null -ne $pool) {
+        $module.Result.exists = $true
+    }
+    $appPoolInfoDict = @{
+        name = $name
+        info = @{}
+    }
+
+    $elements = @("attributes", "cpu", "failure", "processModel", "recycling")
```
Comment (jborean93) [inline plugins/modules/web_app_pool_info.ps1:42]: Is there a reason why `attributes` is separate? We try and make the return value structure similar to the non info module so we should have `$appPoolInfoDict = @{ name = $name; attributes = @{} }` and the attributes returned in a similar format to how people would specify `microsoft.iis.web_app_pool: attribute=...`. Keeping them in a similar format helps users to understand how they can set the attribute based on the info returned either for an the same pool or another one.

File: plugins/modules/web_app_pool_info.yml (hunk 18 above)
Comment (ronger4) [reply plugins/modules/web_app_pool_info.yml:34]: Just for clarification, you suggest me to remove the 'name' key in the example or remove this example all together?
I added this example just to specify the user that if he wants to get all the pools he can.

Comment (jborean93) [reply plugins/modules/web_app_pool_info.yml:34]: Yep make it just

```yaml
  - name: Returns information about all application pools that exist on the system
    microsoft.iis.web_app_pool_info:
    register: stored_info_all
```

File: plugins/modules/web_app_pool_info.ps1 (hunk 19 above)
Comment (ronger4) [reply plugins/modules/web_app_pool_info.ps1:42]: To be honest the reason for that is that I aligned with the way it was implemented originally.
I don't see an issue to change it the way to suggest,
Just to be clear, the fixed format print will be something as follows:
{
    "name": "name1"
    "attributes": {
        "CLRConfigFile": "",
        "autoStart": true,
        "enable32BitAppOnWin64": false,
        "enableConfigurationOverride": true,
        "managedPipelineMode": "Classic",
        "managedRuntimeLoader": "webengine4.dll",
        ....
        ....
 
        "cpu": {
            ......
            ....
        }
        "failure": {
        .....
        ....
        },
        .......
    }
   
   
   
   So basically, include keys like cpu, failure , processModel, etc.. under attributes too.

Comment (ronger4) [reply plugins/modules/web_app_pool_info.ps1:42]: Implemented in
https://github.com/ansible-collections/microsoft.iis/commit/49e95c9684ce9ce31dba2a03438eda006795c927

File: plugins/modules/web_app_pool_info.yml (hunk 18 above)
Comment (ronger4) [reply plugins/modules/web_app_pool_info.yml]: https://github.com/ansible-collections/microsoft.iis/commit/49e95c9684ce9ce31dba2a03438eda006795c927

File: plugins/modules/web_app_pool_info.ps1 (hunk 17 above)
Comment (ronger4) [reply plugins/modules/web_app_pool_info.ps1]: https://github.com/ansible-collections/microsoft.iis/commit/49e95c9684ce9ce31dba2a03438eda006795c927

**Review by jborean93 — APPROVED** [review]
Let's get this merged in so other PRs who rely on this module can run. I think we need to look into the `attribute` handling as currently `web_app_pool` and `web_app_pool_info` differ in how they handle them. The `web_app_pool` module flattens the dictionary path so a value is specified through `processModel.identityType: value` whereas `web_app_pool_info` returns it as

```yaml
attributes:
  processModel:
    identityType: value
```

The module and the `_info` output should align with each other so it can be clear how a user specifies such an option. We either need to update `_info` to return a flatten key or the normal module to accept a dict rather than flattened keys.

I think the latter makes more sense as flattening the return keys in `_info` makes it harder to access them in Ansible. It should also hopefully make the code easier to maintain when accessing each attribute element to set.
